dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -0,0 +1,212 @@
+    public Supplier<Channel> get() {
+        return new StickySessionSupplier(delegate);
+    }
+
+    @Override
+    public String toString() {
+        return "StickyEndpointChannels2{}";
+    }
+
+    private static final class StickySessionSupplier implements Supplier<Channel> {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router = new DefaultStickyRouter();
+
+        private StickySessionSupplier(EndpointChannelFactory channelFactory) {
+            this.channelFactory = channelFactory;
+        }
+
+        @Override
+        public Channel get() {
+            return new Sticky(channelFactory, router);
+        }
+    }
+
+    private static final class Sticky implements EndpointChannelFactory, Channel {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router;
+        private final RoutingKey routingKey;
+
+        private Sticky(EndpointChannelFactory channelFactory, StickyRouter router) {
+            this.router = router;
+            this.channelFactory = channelFactory;
+            this.routingKey = RoutingKey.create();
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            return new StickyEndpointChannel(router, channelFactory.endpoint(endpoint), routingKey);
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+            return endpoint(endpoint).execute(request);
+        }
+
+        @Override
+        public String toString() {
+            return "Sticky{" + channelFactory + '}';
+        }
+    }
+
+    public static StickyEndpointChannels2 create(EndpointChannelFactory endpointChannelFactory) {
+        return new StickyEndpointChannels2(endpointChannelFactory);
+    }
+
+    interface StickyRouter {
+        ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel);
+    }
+
+    private static final class DefaultStickyRouter implements StickyRouter {
+
+        private final InFlightCallCallback callback = new InFlightCallCallback();
+
+        @Nullable
+        private volatile HostId hostId;
+
+        @Nullable
+        private volatile ListenableFuture<Response> callInFlight;
+
+        @Override
+        public ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel) {
+            if (hostId != null) {
+                return executeWithHostId(hostId, request, endpointChannel);
+            }
+
+            synchronized (this) {
+                if (hostId != null) {
+                    request.attachments().put(RoutingAttachments.EXECUTE_ON_HOST_ID_KEY, hostId);
+                    return endpointChannel.execute(request);
+                }
+
+                if (callInFlight == null) {
```

Comment:
I find these easier to read if we do a single volatile read at the top before our conditional
```suggestion
                ListenableFuture<Response> callInFlightSnapshot = callInFlight;
                if (callInFlightSnapshot == null) {
```

Reply:
Same goes for the `hostId` reference as well. It can be refreshed within the synchronized block.